content/blog/multiplayer/index.md

```diff
@@ -0,0 +1,93 @@
+
+The problem was that Excalidraw was built to be a fundamentally single-player experience. However, we ([@idlewinn](https://twitter.com/edwinlin1987) and [@floydophone](https://twitter.com/floydophone)) were able to quickly add basic multiplayer support over a long weekend, and are continuing to make improvements to it every day.
+
+## The requirements
+
+- We wanted to be able to use this in our day jobs. In order to do that, it had to support [end-to-end encryption](/end-to-end-encryption/).
+- We did not want to store anything server-side. Therefore, we opted for a pseudo-P2P model, where a central server relays end-to-end encrypted messages to all the peers in the room, but does no centralized coordination.
+
+## Excalidraw's single-player architecture
+
+Originally, Excalidraw kept an array of all the different drawn shapes -- called `ExcalidrawElement`s -- in Z-index order. The `ExcalidrawElement` interface looked something like this:
+
+```typescript
+interface ExcalidrawElement {
+  id: string;
+  type: "square" | "circle" | "arrow";
+  width: number;
+  height: number;
+  // ... other fields describing the shape ...
+  canvas: HTMLCanvasElement;
+  isSelected: boolean;
+}
+```
+
+This architecture was very easy to use client-side, but as we'll see, presented some challenges as we moved to multiplayer.
+
+## Moving to multiplayer
+
+We actually had two parallel multiplayer implementations. One was built on top of [Firebase](https://firebase.com/), the other on top of [Socket.io](https://socket.io/). We ended up going with Socket.io, but both implementations developed the same set of bugs, and the mitigations we applied in one branch were easily ported to the other.
+
+Conceptually, we want to share the array of `ExcalidrawElement`s amongst multiple players. The first step was to refactor `ExcalidrawElement` to remove any state specific to a single player. This meant pulling out `canvas` and `isSelected` properties and storing them in a different data structure, as an `HTMLCanvasElement` can't be shared over the network, and every player will have a different selection state. Once we had these properties moved out, we started sharing the `ExcalidrawElement` array between peers.
+
+## Socket.IO
+
+When we started investigating multiplayer, we knew that WebSockets would be an important part of the this project because they enable rapid message based communication between clients without polling the server - in order to provide a smooth user experience we would need to sync very frequently. Socket.IO was a great candidate for this project because it's a popular open source library with a lot of support, it has many features relevant to our use case right out of the box, and it has some level of support even for browsers that don't have support for the [WebSocket API](https://developer.mozilla.org/en-US/docs/Web/API/WebSockets_API).
```

Comment:
I'd maybe drop the `it has some level of support even for browsers that don't have support for the WebSocket API` part because this is in some way or another supported even in prehistoric browsers, and we also support only new version.